extensions/cms/src/cmsUtils.ts

```diff
@@ -70,49 +105,72 @@ export class CmsUtils {
-		return this.getCmsService().then((service) => {
-			return service.getRegisteredServers(ownerUri, relativePath).then((result) => {
-				if (result && result.registeredServersList && result.registeredServersList) {
-					return result;
-				}
-			});
-		});
+		const cmsService = await this.getCmsService();
+		const result = await cmsService.getRegisteredServers(ownerUri, relativePath);
+		if (result && result.registeredServersList && result.registeredServersList) {
+			return result;
+		}
+	}
+
+	public static didConnectionChange(connectionA: azdata.connection.Connection, connectionB: azdata.connection.Connection): boolean {
+		return (connectionA !== connectionB) || ((connectionA.connectionId === connectionB.connectionId) &&
+			(connectionA.options.savePassword !== connectionA.options.savePassword));
 	}
 
 	public async createCmsServer(connection: azdata.connection.Connection,
-		name: string, description: string): Promise<mssql.ListRegisteredServersResult> {
+		name: string, description: string): Promise<CreateCmsResult> {
 		let provider = await this.getCmsService();
 		connection.providerName = connection.providerName === cmsProvider ? mssqlProvider : connection.providerName;
 		let ownerUri = await azdata.connection.getUriForConnection(connection.connectionId);
 		if (!ownerUri) {
 			// Make a connection if it's not already connected
-			await azdata.connection.connect(Utils.toConnectionProfile(connection), false, false).then(async (result) => {
-				ownerUri = await azdata.connection.getUriForConnection(result.connectionId);
-			});
+			let initialConnectionProfile = CmsUtils.getConnectionProfile(connection);
+			let result = await azdata.connection.connect(initialConnectionProfile, false, false);
+			ownerUri = await azdata.connection.getUriForConnection(result.connectionId);
+			// If the ownerUri is still undefined, then open a connection dialog with the connection
+			if (!ownerUri) {
+				await this.connection(initialConnectionProfile).then(async (result) => {
+					if (result) {
+						ownerUri = await azdata.connection.getUriForConnection(result.connectionId);
+						connection = result;
+					}
+				}, (error) => {
+					// cancel pressed on connection dialog
+					const errorText = localize('cms.error.cancelConnectionDialog', 'The server is disconnected. Please connect to the server to continue.');
+					return Promise.reject(error ? new Error(error.message) : errorText);
+				});
+			}
 		}
 		return provider.createCmsServer(name, description, connection, ownerUri).then((result) => {
 			if (result) {
-				return Promise.resolve(result);
-			} else {
-				return Promise.reject(null);
+				const createCmsResult: CreateCmsResult = {
+					listRegisteredServersResult: result,
+					connection: connection
+				};
+				return Promise.resolve(createCmsResult);
 			}
+		}, (error) => {
+			return Promise.reject(error.message);
 		});
```

Comment:
You shouldn't be rejecting with just a string, it should always be an Error. But also you're not anything here anyways so there's no need for this handler (the error will just bubble up regardless).

It'd be great to also just refactor this into :

const result: ListRegisteredServersResult = await provider.createCmsServer(name, description, connection, ownerUri);
if(result) { 
     const createCmsResult: CreateCmsResult = {
     listRegisteredServersResult: result,
     connection: connection
    };
    return createCmsResult;
} else {
     throw new Error(...) // does an undefined result even make sense here? I would expect the function to either return a non-undefined result OR just throw
}

Reply:
You're still not throwing an error if the result comes back undefined. I would say that it should be reasonable to expect that this call ALWAYS comes back with a non-undefined result or throws an error in which case checking the result for undefined is pointless. I'd suggest verifying if this is the case and then we can discuss if that's not the case. 

---
In reply to: [293573244](https://github.com/microsoft/azuredatastudio/pull/5989#discussion_r293573244) [](ancestors = 293573244)